python/union_find.py

```diff
@@ -0,0 +1,126 @@
+from typing import List, Dict
+
+INITIAL_RANK = 0
+
+
+class Node:
+    """
+    Class implementing a node in an union-find data structure.
+    Stores the current node's rank and a reference to its parent node.
+    """
+
+    def __init__(self, parent_id: int, rank: int = INITIAL_RANK):
+        self._parent = parent_id
+        self._rank = rank
+
+    @property
+    def parent(self) -> int:
+        return self._parent
+
+    @parent.setter
+    def parent(self, x: int):
+        self._parent = x
+
+    @property
+    def rank(self) -> int:
+        return self._rank
+
+    @rank.setter
+    def rank(self, x: int):
+        self._rank = x
+
+
+class UnionFind:
+    """
+    Class implementing a disjoint-set data structure.
+    """
+
+    def __init__(self, node_ids: List[int]):
+        self.nodes: Dict[int, Node] = {node_id: Node(parent_id=node_id, rank=INITIAL_RANK) for node_id in node_ids}
+
+    def parent(self, node_id: int) -> int:
+        """
+        Returns given node's parent's ID.
+
+        :param node_id: Node ID
+        """
+        return self.nodes[node_id].parent
+
+    def grandparent(self, node_id: int) -> int:
+        """
+        Returns given node's grandparent's ID.
+
+        :param node_id: Node ID
+        """
+        return self.parent(self.parent(node_id))
```

Comment:
What happens when there is no grandparent for a certain node?

Reply:
At initialization, each node is assigned with itself as its parent (and conversely its grandparent, see line 55).
Since there are no actions that would clear a node’s `.parent` attribute, at all times every node has a grandparent.